### PR TITLE
feat: Claude CLI flags — settings panel, in-chat surfaces, and invocation banner

### PR DESCRIPTION
**Pull request by codefriar**
## Summary

Three connected features around exposing and visualizing the `claude` CLI invocation Claudette spawns on every turn:

- **Claude CLI flags settings panel** — discovers `claude --help` at app boot, lets users toggle any non-reserved flag (boolean / value / enum) in a new Settings → Claude flags section. Selections persist globally with per-repo overrides; resolved flags are appended to argv on every turn via a new `extra_claude_flags` field on `AgentSettings`.
- **In-chat visibility** — a `ShieldAlert` badge on chat tabs when `--dangerously-skip-permissions` is on for the workspace's repo, and a `Flag` icon in the composer toolbar whose hover tooltip lists every active extra flag (capped at 10 with overflow). Both consume a per-workspace cache that invalidates when Settings writes a flag value.
- **CLI invocation banner** — the literal `claude` invocation Claudette ran is captured at session start (first emit wins), persisted to a new `chat_sessions.cli_invocation` column, and rendered as the first block of output in every chat tab. Sensitive flag values (`--mcp-config`, `--settings`, `--append-system-prompt`) render as `<redacted>`; the prompt positional renders as `<prompt>`. Works for both the Tauri desktop path and the WebSocket server path (remote workspaces).

Plus a small `scripts/dev.sh` fix: self-bootstraps `mise` and invokes the npm-installed `tauri` binary directly so the launcher works on any contributor's machine without a separate `cargo install tauri-cli`.

## Architecture highlights

- New `claudette::claude_help` parser + `claudette::claude_flags_store` resolver in the library crate. `RESERVED_FLAGS` denylist filters out flags Claudette already manages (`--model`, `--permission-mode`, `--mcp-config`, `--session-id`, …) so users can't accidentally break the agent bridge.
- `StreamEvent::System` gains an optional `command_line` field plus a `system_command_line(line)` constructor; both spawn paths (`run_turn`, `PersistentSession::start`) emit it on the first invocation.
- New `cli_invocation TEXT` column on `chat_sessions` via additive migration `20260507180353_chat_sessions_cli_invocation`. DB write helper enforces first-emit-wins via `WHERE cli_invocation IS NULL`.
- New Tauri command `set_session_cli_invocation` mirrors persistence on the desktop side; `claudette-server`'s event-loop intercepts the `command_line` subtype to mirror persistence on the remote side.
- Per-repo override semantics stored as sentinel-gated keys in `app_settings`: `repo:{repo_id}:claude_flag:{name}:override` distinguishes "explicitly disabled here" from "inheriting from global."

## Test plan

- [ ] **Backend (Rust)**: `cargo test --workspace --all-features` — passes (formatter tests, DB round-trip tests, agent emission tests, plus all existing tests).
- [ ] **Backend lint**: `cargo clippy -p claudette -p claudette-server --all-targets -- -D warnings` — clean.
- [ ] **Backend fmt**: `cargo fmt --all --check` — clean.
- [ ] **Tauri build**: `cargo check -p claudette-tauri --all-targets` — clean (after `bash scripts/stage-cli-sidecar.sh --profile debug` for the sidecar).
- [ ] **Frontend type check**: `cd src/ui && bunx tsc -b` — zero errors.
- [ ] **Frontend tests**: `cd src/ui && bun run test` — all logic tests pass; new banner + slice + flag-row + flag-section tests included.
- [ ] **CSS lint**: `cd src/ui && bun run lint:css` — design-token check passes.
- [ ] **End-to-end (`bash scripts/dev.sh`)**:
  - Toggle `--dangerously-skip-permissions` in Settings → Claude flags. Open a workspace, send a turn. Tab shows the danger badge; composer Flag icon hover lists the flag.
  - Above the message list, the CLI invocation banner renders the redacted `claude --print … --session-id <uuid> <prompt>` line. Sensitive values show `<redacted>`.
  - Reload the app — banner persists. Open a session created before the migration — banner gracefully absent.
  - In `RepoSettings`, override a flag for one repo; confirm the override applies only to that repo's workspaces. Toggle the override off — flag reverts to global.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by codefriar** [comment]
closes #651

**Comment by codefriar** [comment]
Nothing quite so humbling as one's first PR in a new language.

**Comment by doomspork** [comment]
All good @codefriar! Appreciate the contribution 🎉 Happy to take over from here if you'd like. I can get Copilot's feedback incorporated and address any conflicts. 😬

**Comment by jamesbrink** [comment]
> Nothing quite so humbling as one's first PR in a new language.

This is normal for co-pilot.. a review with no feedback is like a unicorn

**Comment by codefriar** [comment]
@doomspork - i'm addressing copilots comments atm.

**Comment by codefriar** [comment]
What else can I do to help move this along?

File: src/agent/session.rs [hunk 1]
```diff
@@ -585,6 +602,7 @@ mod tests {
             mcp_config: Some(r#"{"mcpServers":{}}"#.to_string()),
             disable_1m_context: false,
             hook_bridge: None,
+            extra_claude_flags: Vec::new(),
```
Comment (doomspork) [inline src/agent/session.rs]: 🔴 **Blocker — desktop path silently drops every user-toggled flag.**

This test populates `extra_claude_flags: Vec::new()`, but `build_persistent_args` (line 359 in this file) never reads it — the function emits its argv and returns at line 447 without iterating `settings.extra_claude_flags`. Meanwhile the desktop app spawns turns via `PersistentSession::start` (`src-tauri/src/commands/chat/send.rs:1631`), which calls `build_persistent_args`. So on the dominant user surface, toggling any flag in the new Settings panel does nothing on the next turn.

Only `build_claude_args` got the change (`src/agent/args.rs:200-209`), and that's reached only from `agent::run_turn`, which is called only from the WebSocket server path (`src-server/src/handler.rs:619`). Net: the feature works on remote sessions and is silently broken on every desktop session.

Fix: mirror the loop from `args.rs:200-209` into `build_persistent_args` (no positional prompt to worry about; persistent argv has no trailing prompt), and add a parallel test next to `test_build_persistent_args_all_flags_combined` asserting non-empty `extra_claude_flags` make it into argv.

File: src/ui/src/components/chat/composer/ComposerToolbar.tsx [hunk 2]
```diff
@@ -34,6 +37,20 @@ export function ComposerToolbar({ sessionId, disabled }: ComposerToolbarProps) {
...
+  );
+  const loadWorkspaceClaudeFlags = useAppStore(
+    (s) => s.loadWorkspaceClaudeFlags,
+  );
+
+  useEffect(() => {
+    if (!claudeFlagsState) {
+      void loadWorkspaceClaudeFlags(workspaceId, repoId);
+    }
+  }, [workspaceId, repoId, claudeFlagsState, loadWorkspaceClaudeFlags]);
```
Comment (doomspork) [inline src/ui/src/components/chat/composer/ComposerToolbar.tsx:51]: 🔴 **Duplicate fetch race with `SessionTabs.tsx:830-834`.**

Both this effect and the matching one in `SessionTabs` fire the moment `claudeFlagsByWorkspace[wsId]` is `undefined`. They read the store before either has a chance to write the `loading` placeholder, so both call `loadWorkspaceClaudeFlags`, doubling the Tauri round-trips per workspace switch.

The slice's status field is the natural gate — short-circuit inside `loadWorkspaceClaudeFlags` when an in-flight load already exists for the workspace, or check `claudeFlagsState?.status !== "loading"` before scheduling here.

File: src/ui/src/components/chat/SessionTabs.tsx [hunk 3]
```diff
@@ -805,13 +807,32 @@ function SessionTab({
...
       inputRef.current?.focus();
       inputRef.current?.select();
     }
   }, [editing]);
 
+  useEffect(() => {
+    if (!flagsEntry && repoId) {
+      void loadWorkspaceClaudeFlags(session.workspace_id, repoId);
+    }
+  }, [flagsEntry, repoId, session.workspace_id, loadWorkspaceClaudeFlags]);
```
Comment (doomspork) [inline src/ui/src/components/chat/SessionTabs.tsx:834]: 🔴 Companion to the `ComposerToolbar` race — same effect, same gate. Whichever side gets the deduplication treatment, both should agree.

File: src/ui/src/stores/slices/workspaceClaudeFlagsSlice.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    const current = get().claudeFlagsByWorkspace;
+    if (!(workspaceId in current)) return;
+    const next = { ...current };
+    delete next[workspaceId];
+    set({ claudeFlagsByWorkspace: next });
+  },
+
+  invalidateAllWorkspaceClaudeFlags: () => {
+    set({ claudeFlagsByWorkspace: {} });
+  },
```
Comment (doomspork) [inline src/ui/src/stores/slices/workspaceClaudeFlagsSlice.ts:140]: 🟡 Sledgehammer invalidation. Every Settings toggle wipes `claudeFlagsByWorkspace = {}`, so on the next render every mounted `SessionTab` and `ComposerToolbar` triggers its own re-fetch — N tauri round-trips per toggle, where N = open workspaces.

A targeted invalidation per affected workspace, or a single shared cache keyed by `repoId` rather than per-workspace, would cut this. Not catastrophic (Settings toggles are rare) but easy to fix.

File: src-tauri/src/commands/chat/send.rs [hunk 5]
```diff
@@ -1306,6 +1306,35 @@ pub async fn send_chat_message(
...
+                        );
+                        Vec::new()
+                    }
+                }
+            }
+            crate::state::ClaudeFlagDiscovery::Err(msg) => {
+                eprintln!("[chat] claude flag discovery failed: {msg}");
+                Vec::new()
+            }
+            crate::state::ClaudeFlagDiscovery::Loading => Vec::new(),
```
Comment (doomspork) [inline src-tauri/src/commands/chat/send.rs:1349]: 🟡 Boot-time discovery race. `discover_claude_flags()` is spawned in `setup` but not awaited, so a turn started in the first ~100ms after launch hits this `Loading` arm and silently sends `extra_claude_flags = Vec::new()`. The user's flags are dropped with no UI feedback — bug-by-rare-timing.

`claude --help` returns fast enough that you could just `block_on` it in `setup` (or `await` it before `tauri::Builder::run`). At minimum, log at warn-level when this branch fires so it's traceable in dev tools.

File: src/ui/src/components/settings/sections/ClaudeFlagsSettings.tsx [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+      setCachedDefs(defs);
+    } catch (e) {
+      if (isStillLoading(e)) {
+        // Don't surface a Retry banner during the boot-time discovery
+        // window — re-poll until discovery resolves to Ok or a real error.
+        setTimeout(() => {
+          if (useAppStore.getState().claudeFlagDefs === null) {
+            void loadDefs();
+          }
+        }, 500);
```
Comment (doomspork) [inline src/ui/src/components/settings/sections/ClaudeFlagsSettings.tsx:67]: 🟡 No unmount cancellation. If the user navigates away from Settings during boot-time discovery, this `setTimeout` keeps firing `loadDefs` recursively — which writes to the global Zustand store, so the actual harm is bounded, but it's still a leak. Standard `useEffect` cleanup with a ref-flag (or an `AbortController` if you want to be thorough) would close it.

File: src/agent/args.rs [hunk 7]
```diff
@@ -1,6 +1,111 @@
 use super::AgentSettings;
 use super::types::FileAttachment;
 
+/// Flags whose value is large or sensitive: `--mcp-config` (inline JSON, may
+/// carry tokens), `--settings` (inline JSON, may carry hook bridge socket +
+/// auth token), `--append-system-prompt` (potentially huge user prompt).
+/// Renders these as `<flag> <redacted>` in `format_redacted_invocation`.
```
Comment (doomspork) [inline src/agent/args.rs]: 🟡 Redaction list misses some user-toggleable sensitive fields. The most obvious gap: `--system-prompt` (value-taking, not in `RESERVED_FLAGS`, content is the literal twin of the already-redacted `--append-system-prompt`). Also worth considering: `--agents` (JSON), `--betas`, `--json-schema`.

Minimum fix: add `--system-prompt`. A heuristic alternative would be to redact any flag whose `value_placeholder` contains `"prompt"` or `"json"`, but a hand-curated list is fine if you'd rather stay explicit.

File: src/claude_help.rs [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+    "--print",
+    "--output-format",
+    "--input-format",
+    "--verbose",
+    "--include-partial-messages",
+    "--session-id",
+    "--resume",
+    "--model",
+    "--permission-mode",
+    "--permission-prompt-tool",
```
Comment (doomspork) [inline src/claude_help.rs:46]: 🟡 `RESERVED_FLAGS` is a hand-maintained list with no compiler-enforced link to its emitters in `args.rs` / `session.rs`. If someone adds a new Claudette-managed flag and forgets to add it here, users can toggle a flag that breaks the agent bridge (output framing, session-id ownership, permission-prompt protocol) — and there's no test that fails.

Not a blocker, but a CI grep step ("every quoted `--…` literal in `build_claude_args`/`build_persistent_args` appears in `RESERVED_FLAGS`") would close the gap mechanically. At minimum, the doc comment here should remain a sticky reminder during code review of those two builders.

File: src-server/src/handler.rs [hunk 9]
```diff
@@ -1197,3 +1226,51 @@ async fn handle_archive_chat_session(
...
+/// Resolve the cached `claude --help` defs (running discovery on first
+/// call) into a clone the caller can use synchronously alongside a
+/// `!Send` `Database`. Returning a clone — rather than holding the
+/// `RwLock` guard across the resolver call — keeps the handler future
+/// `Send`, which `tokio::spawn` requires.
+async fn cached_claude_flag_defs(
+    state: &Arc<crate::ws::ServerState>,
+) -> Result<Vec<claudette::claude_help::ClaudeFlagDef>, String> {
+    {
+        let guard = state.claude_flag_defs.read().await;
```
Comment (doomspork) [inline src-server/src/handler.rs:1259]: 🟡 Server-side cache never refreshes. Lazily populated on first turn, retained forever — if the user upgrades `claude` while the server is running, defs go stale with no Retry path (the desktop has `refresh_claude_flags` but the server has no equivalent). Acceptable for v1, worth a TODO at minimum.

**Review by doomspork — DISMISSED** [review]
Thanks for the substantial work here — the architecture is solid, test coverage is genuinely good (pure-helper extraction throughout, banner argv-shape edge cases are well covered), and the i18n is complete across all five locales. Two blockers and a handful of follow-up concerns inline.

**Blockers:**
1. `build_persistent_args` doesn't append `extra_claude_flags`, so the desktop spawn path silently drops every user-toggled flag. (See inline on `src/agent/session.rs`.)
2. Both `SessionTabs` and `ComposerToolbar` race to populate the per-workspace flag cache, doubling the round-trips per workspace switch.

**Follow-ups (non-blocking but worth addressing):**
- `invalidateAllWorkspaceClaudeFlags` nukes every cache on each toggle — N round-trips per Settings change.
- Boot-time discovery race: the first turn after launch silently runs without user flags if `claude --help` hasn't returned yet.
- `setTimeout` polling in `ClaudeFlagsSettings` has no unmount cancellation.
- `REDACTED_VALUE_FLAGS` omits user-toggleable sensitive fields, notably `--system-prompt` (twin of the already-redacted `--append-system-prompt`).
- `RESERVED_FLAGS` has no compiler-enforced link to the actual emitters in `args.rs` / `session.rs` — easy to forget to update.
- Server-side flag-defs cache never refreshes (no `refresh_claude_flags` equivalent on the WebSocket side).

**Strong points:** sentinel-keyed `:override` design (lets per-repo "off" survive a global re-enable), additive migration with `WHERE cli_invocation IS NULL` first-emit-wins, pure-helper extraction for testability, fault-tolerant banner emission (`try_send`, explicit "UX-cosmetic" comments).

File: src/agent/session.rs (hunk 1 above)
Comment (jamesbrink) [reply src/agent/session.rs]: Addressed in 0b97791 — `build_persistent_args` now mirrors the loop in `build_claude_args` and pushes each `(name, value)` from `settings.extra_claude_flags` onto argv. The previously-asserting test was extended into `test_build_persistent_args_extra_claude_flags`, which covers both boolean and value-bearing flags on the persistent path.

File: src/ui/src/components/chat/composer/ComposerToolbar.tsx (hunk 2 above)
Comment (jamesbrink) [reply src/ui/src/components/chat/composer/ComposerToolbar.tsx:51]: Addressed in 77dc015 — `loadWorkspaceClaudeFlags` now short-circuits when the workspace entry is already in `loading` or `ready` status. Whichever sibling (`ComposerToolbar` or `SessionTabs`) fires first wins; the other becomes a no-op until `invalidate*` clears the entry.

File: src/ui/src/components/chat/SessionTabs.tsx (hunk 3 above)
Comment (jamesbrink) [reply src/ui/src/components/chat/SessionTabs.tsx:834]: Same fix as the `ComposerToolbar` thread — 77dc015 gates `loadWorkspaceClaudeFlags` on the slice's status field, so both effects deduplicate against the same in-flight load.

File: src/ui/src/stores/slices/workspaceClaudeFlagsSlice.ts (hunk 4 above)
Comment (jamesbrink) [reply src/ui/src/stores/slices/workspaceClaudeFlagsSlice.ts:140]: Addressed in fb99b37 — added `invalidateClaudeFlagsForRepo(repoId)`. `ClaudeFlagsSettings` calls the targeted variant when `scope.repoId` is set; the global wipe only fires when the user toggles a true global default. Targeted invalidation walks `claudeFlagsByWorkspace` and only evicts entries whose `repoId` matches.

File: src-tauri/src/commands/chat/send.rs (hunk 5 above)
Comment (jamesbrink) [reply src-tauri/src/commands/chat/send.rs:1349]: Addressed in 42874d1 — `claude --help` discovery is now `tauri::async_runtime::block_on`'d inside Tauri `setup`, so the first turn always sees fully-resolved defs. Costs a beat at startup but eliminates the silent-flag-drop window.

File: src/ui/src/components/settings/sections/ClaudeFlagsSettings.tsx (hunk 6 above)
Comment (jamesbrink) [reply src/ui/src/components/settings/sections/ClaudeFlagsSettings.tsx:67]: Addressed in ec66dc1 — added a `useEffect` cleanup that clears `pollTimerRef.current` on unmount, so navigating away from Settings during boot-time discovery no longer leaks the recursive `setTimeout`.

File: src/agent/args.rs (hunk 7 above)
Comment (jamesbrink) [reply src/agent/args.rs]: Addressed in 60f8a74 — `--system-prompt` (plus the other JSON-bearing flags you flagged) is now in `REDACTED_VALUE_FLAGS`. New round-trip assertion in `args.rs` confirms `--system-prompt <redacted>` appears in the formatted invocation string.

File: src/claude_help.rs (hunk 8 above)
Comment (jamesbrink) [reply src/claude_help.rs:46]: Addressed in 7bbf24e — new test in `claude_help.rs` scans the source of `build_claude_args` and `build_persistent_args` for `--…` literals and asserts every one is in `RESERVED_FLAGS` (or a small explicit test allowlist for value tokens). Forgetting to reserve a Claudette-managed flag now fails CI.

File: src-server/src/handler.rs (hunk 9 above)
Comment (jamesbrink) [reply src-server/src/handler.rs:1259]: Addressed in 6d4bff8 — added a `refresh_claude_flags` JSON-RPC method on the WebSocket server handler that nulls the cached `claude_flag_defs`, so the next turn re-runs `discover_claude_flags()`. Mirrors the desktop-side `refresh_claude_flags` Tauri command.

File: src/agent/session.rs [hunk 10]
```diff
@@ -119,6 +131,11 @@ impl PersistentSession {
 
         let (event_tx, _) = tokio::sync::broadcast::channel::<AgentEvent>(2048);
 
+        // Persistent sessions only build argv once — `start` is the right
+        // (and only) spot to capture the invocation. No `is_resume` guard
+        // because the persistent path doesn't re-spawn within a session.
+        emit_invocation_to_broadcast(&event_tx, claude_path.as_os_str(), &args);
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/agent/session.rs]: Addressed in c8d8665b. You're right — `broadcast::Sender::send` drops messages when the receiver count is zero, so the original placement at `PersistentSession::start` was effectively a no-op for the desktop spawn path.

The fix:
- Capture the formatted invocation line at start time (still the right spot since argv is built once per session) and store it on the session.
- Replay it inside `send_turn` after the first `subscribe()` call, gated by an `AtomicBool` so subsequent turns skip the re-broadcast.
- Removed the now-unused `emit_invocation_to_broadcast` helper.

Both the DB write helper and the UI's `applyCommandLineEvent` are first-emit-wins, so a stray re-emit would be cosmetic — the gate is purely an optimization.

File: src/ui/src/components/settings/sections/ClaudeFlagsSettings.tsx [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+    setDefsError(null);
+    try {
+      const defs = await listClaudeFlags();
+      setCachedDefs(defs);
+    } catch (e) {
+      if (isStillLoading(e)) {
+        // Don't surface a Retry banner during the boot-time discovery
+        // window — re-poll until discovery resolves to Ok or a real error.
+        if (pollTimerRef.current) clearTimeout(pollTimerRef.current);
+        pollTimerRef.current = setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/components/settings/sections/ClaudeFlagsSettings.tsx:63]: Addressed in 9954da1a. Removed the `finally` block and moved `setDefsLoading(false)` into the success and real-error branches explicitly. The polling branch now keeps `defsLoading` true through the entire re-poll window, so the `claude_flags_loading` hint stays visible until `listClaudeFlags` resolves to either a defs payload or a non-still-loading error.

File: src/ui/src/components/chat/composer/ClaudeFlagsTooltip.tsx [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+                  <span>{flag.name}</span>
+                  {flag.value !== undefined && flag.value !== "" && (
+                    <span className={styles.value}>{flag.value}</span>
+                  )}
+                </li>
+              ))}
+            </ul>
+            {hidden > 0 && (
+              <div className={styles.more}>…and {hidden} more</div>
+            )}
```
[the remark this answers is not in the document]
Comment (jamesbrink) [reply src/ui/src/components/chat/composer/ClaudeFlagsTooltip.tsx:62]: Addressed in 29cf741c. Added `claude_flags_tooltip_overflow` to all five locale files (en, es, ja, pt-BR, zh-CN) using i18next's `{{count}}` interpolation, and swapped the inline JSX for a `t()` call. Per-locale strings:
- en: `…and {{count}} more`
- es: `…y {{count}} más`
- ja: `…他 {{count}} 件`
- pt-BR: `…e mais {{count}}`
- zh-CN: `…还有 {{count}} 项`
